File: directory_analyzer/rust-ver/src/analyze.rs
use std::{
    collections::HashMap,
    fs::{self, DirEntry},
    path::{self, PathBuf},
};

use crate::{AnalyzedInfo, Args, FastPath, FileTypeInfo, FileTypeInfoRecords, SymlinkInfo, Timer};
use anyhow::Result;

pub fn analyze(args: &Args) -> Result<AnalyzedInfo> {
    let mut out = set_up_anaylzed_info(args);
    search_dirs(args, &mut out)?;
    out.calculate_percentages_for_info();
    if let Some(sym) = out.found_symlinks() {
        //Sanity check to make sure things add up
        debug_assert_eq!(
            sym.found_symlinks(),
            sym.dir_symlinks() + sym.file_symlinks()
        );
    }
    Ok(out)
}

fn search_dirs(args: &Args, analyed_info: &mut AnalyzedInfo) -> Result<()> {
    let mut dirs_to_analyze = vec![args.path_to_analyze().clone()];
    //used to prevent repeatedly counting the same item multiple times while following symlinks
    let mut found_items = Vec::new();
    let mut timer = args.updates().map(Timer::new);
    let ignore_these = set_up_ignore_these(args)?;
    while let Some(dir) = dirs_to_analyze.pop() {
        let cur_dir = dir.read_dir()?;
        for entry in cur_dir {
            let entry = entry?;
            let metadata = entry.metadata()?;
            let path = path::absolute(entry.path())?;
            if let Some(ignore_these) = &ignore_these {
                if ignore_these
                    .iter()
                    .any(|p| path::absolute(p).unwrap() == path)
                {
                    continue;
                }
            }
            if metadata.is_dir() && !found_items.contains(&FastPath::new(&path)) {
                handle_dirs(args, &mut dirs_to_analyze, &entry, analyed_info)?;
                //No need to track found items since there is no risk of loops or double counting
                if args.follow_symlinks() {
                    found_items.push(FastPath::new(&path));
                }
            } else if metadata.is_file() && !found_items.contains(&FastPath::new(&path)) {
                handle_files(args, analyed_info, &entry, &metadata)?;
                //No need to track found items since there is no risk of loops or double counting
                if args.follow_symlinks() {
                    found_items.push(FastPath::new(&path));
                }
            } else if (args.count_symlinks() || args.follow_symlinks())
                && metadata.is_symlink()
                && !found_items.contains(&FastPath::new(&path))
            {
                handle_symlinks(
                    entry,
                    &mut found_items,
                    args,
                    &mut dirs_to_analyze,
                    analyed_info,
                )?;
            }
        }
        if let Some(timer) = &mut timer {
            timer.update();
            if timer.ended() {
                println!(
                    "Update: {} found dirs, {} found files",
                    analyed_info.found_dirs(),
                    analyed_info.found_files()
                );
                timer.reset();
            }
        }
    }

    fn handle_dirs(
        args: &Args,
        dirs_to_analyze: &mut Vec<PathBuf>,
        entry: &DirEntry,
        analyed_info: &mut AnalyzedInfo,
    ) -> anyhow::Result<()> {
        handle_args(args, dirs_to_analyze, entry)?;
        *analyed_info.found_dirs_mut() += 1;

        fn handle_args(
            args: &Args,
            dirs_to_analyze: &mut Vec<PathBuf>,
            entry: &DirEntry,
        ) -> anyhow::Result<()> {
            if !args.no_recurse() {
                dirs_to_analyze.push(entry.path());
            }
            if args.verbose() {
                if args.full_path() {
                    println!("dir: {}", path::absolute(entry.path())?.to_string_lossy());
                } else {
                    println!("dir: {}", entry.path().to_string_lossy());
                }
            }
            Ok(())
        }
        Ok(())
    }

    fn handle_files(
        args: &Args,
        analyed_info: &mut AnalyzedInfo,
        entry: &DirEntry,
        metadata: &fs::Metadata,
    ) -> anyhow::Result<()> {
        handle_file_args(args, analyed_info.file_info_mut(), entry, metadata)?;
        *analyed_info.found_files_mut() += 1;
        *analyed_info.total_bytes_mut() += metadata.len();

        fn handle_file_args(
            args: &Args,
            map: Option<&mut HashMap<String, FileTypeInfo>>,
            entry: &DirEntry,
            metadata: &fs::Metadata,
        ) -> anyhow::Result<()> {
            if args.verbose() {
                if args.full_path() {
                    println!("file: {}", path::absolute(entry.path())?.to_string_lossy())
                } else {
                    println!("file: {}", entry.path().to_string_lossy())
                }
            }

            if let Some(map) = map {
                match entry.path().extension() {
                    Some(ext) => {
                        let ext = ext.to_os_string().to_string_lossy().to_string();
                        add_file_info_to_map(args, ext, map, entry, metadata)?;
                    }
                    //Still want to keep info about files without extensions
                    None => {
                        let ext = "".to_string();
                        add_file_info_to_map(args, ext, map, entry, metadata)?;
                    }
                }
            }
            Ok(())
        }
        Ok(())
    }

    ///Traverses and counts symlinks and if the target is not already counted counts it
    fn handle_symlinks(
        entry: DirEntry,
        found_items: &mut Vec<FastPath>,
        args: &Args,
        dirs_to_analyze: &mut Vec<PathBuf>,
        analyed_info: &mut AnalyzedInfo,
    ) -> Result<(), anyhow::Error> {
        let path = fs::read_link(entry.path())?;
        let metadata = path.metadata()?;
        if args.follow_symlinks() {
            //don't look at entries that have been seen before
            //prevents following symlink loops and counting entries multiple times
            if !found_items.contains(&FastPath::new(&path)) {
                if metadata.is_dir() {
                    handle_dirs(args, dirs_to_analyze, &entry, analyed_info)?;
                } else if metadata.is_file() {
                    handle_files(args, analyed_info, &entry, &metadata)?;
                }

                found_items.push(FastPath::new(&path));
            }
        }

        if args.count_symlinks() {
            //Count the found symlinks here since a symlink can be new but point to a entry that
            //has already been seen before. We still want to count the symlink as found though
            if let Some(symlink) = analyed_info.found_symlinks_mut() {
                if metadata.is_dir() {
                    *symlink.dir_symlinks_mut() += 1;
                } else if metadata.is_file() {
                    *symlink.file_symlinks_mut() += 1;
                }
                *symlink.found_symlinks_mut() += 1;
            }
        }

        Ok(())
    }

    Ok(())
}

fn set_up_ignore_these(args: &Args) -> Result<Option<Vec<PathBuf>>, anyhow::Error> {
    let ignore_these: Option<Vec<_>> = if let Some(s) = args.ignore_entries() {
        let paths: Vec<_> = s.split(',').map(|s| s.trim()).map(PathBuf::from).collect();
        for p in &paths {
            let exists = p.try_exists()?;
            if !exists {
                eprintln!(
                    "WARNING: Can't ignore \"{}\" because it doesn't exist",
                    p.to_string_lossy()
                );
            }
        }
        Some(paths)
    } else {
        None
    };
    Ok(ignore_these)
}

fn add_file_info_to_map(
    args: &Args,
    extension: String,
    map: &mut HashMap<String, FileTypeInfo>,
    entry: &DirEntry,
    metadata: &std::fs::Metadata,
) -> anyhow::Result<()> {
    let t = map.entry(extension).or_insert(FileTypeInfo::new(
        0,
        0,
        FileTypeInfoRecords::default(),
        FileTypeInfoRecords::new(PathBuf::default(), u64::MAX),
    ));
    *t.num_files_mut() += 1;
    *t.size_in_bytes_mut() += metadata.len();
    let path = if args.full_path() {
        path::absolute(entry.path())?
    } else {
        entry.path()
    };
    if metadata.len() > t.largest_file().size() {
        t.set_largest_file(FileTypeInfoRecords::new(path.clone(), metadata.len()));
    }
    if metadata.len() < t.smallest_file().size() {
        t.set_smallest_file(FileTypeInfoRecords::new(path, metadata.len()));
    }
    Ok(())
}

///Configures `AnalyzedInfo` based on the `Args` given
fn set_up_anaylzed_info(args: &Args) -> AnalyzedInfo {
    let mut out = AnalyzedInfo::default();
    if args.file_info() {
        out.set_file_info(Some(HashMap::default()));
    }
    if args.count_symlinks() {
        out.set_found_symlinks(Some(SymlinkInfo::default()));
    }
    out
}

#[cfg(test)]
mod tests {
    use std::{path::PathBuf, str::FromStr};

    use crate::FileExtension;

    use super::*;

    const TEST_DIR: &str = "../test/";

    #[test]
    fn analyze_default_settings() {
        let test_args = Args::new(
            PathBuf::from_str(TEST_DIR).unwrap(),
            false,
            false,
            false,
            false,
            None,
            false,
            None,
            None,
            true,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(4, 7, Some(SymlinkInfo::new(2, 1, 1)), None, 432);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_no_recurse() {
        let test_args = Args::new(
            PathBuf::from_str(TEST_DIR).unwrap(),
            true,
            false,
            false,
            false,
            None,
            false,
            None,
            None,
            false,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(2, 4, None, None, 407);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_file_info() {
        let test_args = Args::new(
            PathBuf::from_str(TEST_DIR).unwrap(),
            false,
            true,
            false,
            false,
            None,
            false,
            None,
            None,
            false,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let mut hash_map: HashMap<FileExtension, FileTypeInfo> = HashMap::new();
        //Byte values are from windows properties Size: field for each file

        hash_map.insert(
            "txt".to_string(),
            FileTypeInfo::new(
                9,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            ),
        ); //file1.txt
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                14,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //file2.txt
        *h.size_in_bytes_mut() += 14;
        *h.num_files_mut() += 1;
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("file2.txt");
        h.set_largest_file(FileTypeInfoRecords::new(path, 14));
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                9,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //folder1/file5.txt
        *h.size_in_bytes_mut() += 9;
        *h.num_files_mut() += 1;
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                7,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //folder2/file6.txt
        *h.size_in_bytes_mut() += 7;
        *h.num_files_mut() += 1;
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("folder2/file6.txt");
        h.set_smallest_file(FileTypeInfoRecords::new(path, 7));
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                9,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //folder2/folder3/folder4/deepfile1.txt
        *h.size_in_bytes_mut() += 9;
        *h.num_files_mut() += 1;
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("file3.rtf");
        hash_map.insert(
            "rtf".to_string(),
            FileTypeInfo::new(
                196,
                1,
                FileTypeInfoRecords::new(path.clone(), 196),
                FileTypeInfoRecords::new(path, 196),
            ),
        ); //file3.rtf
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("file4.zip");
        hash_map.insert(
            "zip".to_string(),
            FileTypeInfo::new(
                188,
                1,
                FileTypeInfoRecords::new(path.clone(), 188),
                FileTypeInfoRecords::new(path.clone(), 188),
            ),
        ); //file4.zip

        let (total_files, total_bytes) = get_total_files_and_bytes_from_map(&hash_map);
        for (_, info) in hash_map.iter_mut() {
            info.calculate_percentages(total_bytes, total_files);
        }
        let expected = AnalyzedInfo::new(4, 7, None, Some(hash_map), 432);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_follow_symlinks() {
        let test_args = Args::new(
            PathBuf::from_str(TEST_DIR).unwrap(),
            false,
            false,
            true,
            false,
            None,
            false,
            None,
            None,
            true,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(6, 8, Some(SymlinkInfo::new(2, 1, 1)), None, 432);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_ignore_entries() {
        let path = PathBuf::from_str(TEST_DIR).unwrap();
        let mut str1 = TEST_DIR.to_string();
        str1.push_str("folder2/folder3");
        let mut str2 = TEST_DIR.to_string();
        str2.push_str("file4.zip");
        let ignore_entries = format!("{str1}, {str2}");
        let test_args = Args::new(
            path,
            false,
            false,
            false,
            false,
            None,
            false,
            Some(ignore_entries),
            None,
            false,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(2, 5, None, None, 235);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_ignore_entries_file_info() {
        let path = PathBuf::from_str(TEST_DIR).unwrap();
        let mut str1 = TEST_DIR.to_string();
        str1.push_str("folder2/folder3");
        let mut str2 = TEST_DIR.to_string();
        str2.push_str("file4.zip");
        let ignore_entries = format!("{str1}, {str2}");
        let test_args = Args::new(
            path,
            false,
            true,
            false,
            false,
            None,
            false,
            Some(ignore_entries),
            None,
            false,
            false,
        );
        let mut hash_map: HashMap<FileExtension, FileTypeInfo> = HashMap::new();
        //Byte values are from windows properties Size: field for each file

        hash_map.insert(
            "txt".to_string(),
            FileTypeInfo::new(
                9,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            ),
        ); //file1.txt
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                14,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //file2.txt
        *h.size_in_bytes_mut() += 14;
        *h.num_files_mut() += 1;
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("file2.txt");
        h.set_largest_file(FileTypeInfoRecords::new(path, 14));
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                9,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //folder1/file5.txt
        *h.size_in_bytes_mut() += 9;
        *h.num_files_mut() += 1;
        let h = hash_map
            .entry("txt".to_string())
            .or_insert(FileTypeInfo::new(
                7,
                1,
                FileTypeInfoRecords::default(),
                FileTypeInfoRecords::default(),
            )); //folder2/file6.txt
        *h.size_in_bytes_mut() += 7;
        *h.num_files_mut() += 1;
        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("folder2/file6.txt");
        h.set_smallest_file(FileTypeInfoRecords::new(path, 7));

        let mut path = PathBuf::from_str(TEST_DIR).unwrap();
        path.push("file3.rtf");
        hash_map.insert(
            "rtf".to_string(),
            FileTypeInfo::new(
                196,
                1,
                FileTypeInfoRecords::new(path.clone(), 196),
                FileTypeInfoRecords::new(path, 196),
            ),
        ); //file3.rtf

        let (total_files, total_bytes) = get_total_files_and_bytes_from_map(&hash_map);
        for (_, info) in hash_map.iter_mut() {
            info.calculate_percentages(total_bytes, total_files);
        }
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(2, 5, None, Some(hash_map), 235);
        assert_eq!(res, expected);
    }

    fn get_total_files_and_bytes_from_map(hash_map: &HashMap<String, FileTypeInfo>) -> (u32, u64) {
        let total_files = hash_map
            .iter()
            .fold(0, |acc, (_, info)| acc + info.num_files());
        let total_bytes = hash_map
            .iter()
            .fold(0, |acc, (_, info)| acc + info.size_in_bytes());
        (total_files, total_bytes)
    }

    #[test]
    fn analyze_symlink_loop() {
        let test_args = Args::new(
            PathBuf::from_str("../test_symlink_loop/").unwrap(),
            false,
            false,
            true,
            false,
            None,
            false,
            None,
            None,
            true,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(2, 2, Some(SymlinkInfo::new(2, 0, 2)), None, 21);
        assert_eq!(res, expected);
    }

    #[test]
    fn analyze_follow_symlinks_no_count_symlinks() {
        let test_args = Args::new(
            PathBuf::from_str(TEST_DIR).unwrap(),
            false,
            false,
            true,
            false,
            None,
            false,
            None,
            None,
            false,
            false,
        );
        let res = analyze(&test_args).unwrap();
        let expected = AnalyzedInfo::new(6, 8, None, None, 432);
        assert_eq!(res, expected);
    }
}

File: directory_analyzer/csharp-ver/dira/analyze.cs
using System.Diagnostics;
using System.Timers;
using dira.types;

namespace dira
{

    public static class A
    {

        public static AnalyzedInfo Analyze(Args args)
        {
            AnalyzedInfo res = SetUpAnalyzedInfo(args);
            SearchDir(args, ref res);
            res.CalculatePercentagesForInfo();
            return res;
        }

        private static void SearchDir(Args args, ref AnalyzedInfo res)
        {
            var dirs_to_analyze = new Stack<string>();
            dirs_to_analyze.Push(args.PathToAnalyze);
            //used to prevent repeatedly counting the same item multiple times while following symlinks
            var found_items = new List<string>();
            var timer = SetUpTimer(args);
            var ignore_these = SetUpIgnoreThese(args);

            while (dirs_to_analyze.TryPop(out var dir))
            {
                var cur_dir = Directory.EnumerateFileSystemEntries(dir);
                foreach (var entry in cur_dir)
                {
                    var path = Path.GetFullPath(entry).TrimEnd(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
                    var metadata = File.GetAttributes(path);
                    var info = new FileInfo(path);
                    if (ignore_these != null)
                    {
                        if (ignore_these.Any(p => Path.GetFullPath(p).TrimEnd(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar) == path))
                        {
                            continue;
                        }
                    }
                    if (metadata.HasFlag(FileAttributes.Directory) && !found_items.Contains(path) && info.LinkTarget == null)
                    {
                        //Directories
                        HandleDirs(args, ref dirs_to_analyze, entry, ref res);
                        if (args.FollowSymlinks)
                        {
                            found_items.Add(path);
                        }
                    }
                    else if (File.Exists(path) && !found_items.Contains(path) && info.LinkTarget == null)
                    {
                        //Files
                        HandleFiles(args, ref res, entry, info);
                        if (args.FollowSymlinks)
                        {
                            found_items.Add(path);
                        }
                    }
                    else if ((args.CountSymlinks || args.FollowSymlinks) && info.LinkTarget != null && !found_items.Contains(path))
                    {
                        //Symbolic links
                        HandleSymlinks(args, ref found_items, entry, ref dirs_to_analyze, ref res, info);
                    }

                }

                if (timer != null)
                {
                    if ((ulong)timer.Elapsed.Seconds > args.Updates)
                    {
                        timer.Reset();
                        PrintUpdate(res.FoundDirs, res.FoundFiles);
                        timer.Start();
                    }
                }
            }

            static void PrintUpdate(uint found_dirs, uint found_files)
            {
                Console.WriteLine("Update: {0} found dirs, {1} found files", found_dirs, found_files);
            }
            static void HandleDirs(Args args, ref Stack<string> dirs_to_analyze, string entry, ref AnalyzedInfo analyzed_info)
            {
                HandleArgs(args, ref dirs_to_analyze, entry);
                analyzed_info.FoundDirs += 1;

                static void HandleArgs(Args args, ref Stack<string> dirs_to_analyze, string entry)
                {
                    if (!args.NoRecurse)
                    {
                        dirs_to_analyze.Push(entry);
                    }
                    if (args.Verbose)
                    {
                        if (args.FullPath)
                        {
                            Console.WriteLine("dir: {0}", Path.GetFullPath(entry).TrimEnd(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar));
                        }
                        else
                        {
                            Console.WriteLine("dir:{0}", entry);
                        }
                    }
                }
            }

            static void HandleFiles(Args args, ref AnalyzedInfo analyzed_info, string entry, FileInfo info)
            {
                HandleArgs(args, ref analyzed_info, entry, info);
                analyzed_info.FoundFiles += 1;
                analyzed_info.TotalBytes += (ulong)info.Length;

                static void HandleArgs(Args args, ref AnalyzedInfo analyzed_info, string entry, FileInfo info)
                {
                    if (args.Verbose)
                    {
                        if (args.FullPath)
                        {
                            Console.WriteLine("file:{0}", Path.GetFullPath(entry));
                        }
                        else
                        {
                            Console.WriteLine("file: {0}", entry);
                        }
                    }

                    if (analyzed_info.FileInfo != null)
                    {
                        switch (Path.GetExtension(entry))
                        {
                            case string ext when ext != "":
                                //Remove leaving . for consistancy with the rust version
                                AddFileInfoToMap(args, new FileExtension(ext[1..]), ref analyzed_info, entry, info);
                                break;
                            case "":
                                //Still want to keep info about files without extensions
                                AddFileInfoToMap(args, new FileExtension(""), ref analyzed_info, entry, info);
                                break;
                        }
                    }
                }
            }

            static void HandleSymlinks(Args args, ref List<string> found_items, string entry, ref Stack<string> dirs_to_analyze, ref AnalyzedInfo analyzed_info, FileInfo f_info)
            {
                Trace.Assert(f_info.LinkTarget != null);
                var path = Path.GetFullPath(f_info.LinkTarget).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
                var metadata = File.GetAttributes(path);
                var info = new FileInfo(path);
                if (args.FollowSymlinks)
                {
                    //don't look at entries that have been seen before
                    //prevents following symlink loops and counting entries multiple times
                    if (!found_items.Contains(path))
                    {
                        if (metadata.HasFlag(FileAttributes.Directory))
                        {
                            HandleDirs(args, ref dirs_to_analyze, entry, ref analyzed_info);
                        }
                        else if (File.Exists(path))
                        {
                            HandleFiles(args, ref analyzed_info, entry, info);
                        }
                        found_items.Add(path);
                    }
                }
                if (args.CountSymlinks && analyzed_info.FoundSymlinks is SymlinkInfo symlink)
                {
                    //Count the found symlinks here since a symlink can be new but point to a entry that
                    //has already been seen before. We still want to count the symlink as found though
                    if (metadata.HasFlag(FileAttributes.Directory))
                    {
                        var sym = new SymlinkInfo();
                        sym.SetDirSymlinks(symlink.GetDirSymlinks() + 1);
                        sym.SetFileSymlinks(symlink.GetFileSymlinks());
                        analyzed_info.FoundSymlinks = sym;
                    }
                    else if (File.Exists(path))
                    {
                        var sym = new SymlinkInfo();
                        sym.SetDirSymlinks(symlink.GetDirSymlinks());
                        sym.SetFileSymlinks(symlink.GetFileSymlinks() + 1);
                        analyzed_info.FoundSymlinks = sym;
                    }

                }

            }
        }

        /// <summary>
        /// Must make sure AnalyzedInfo.FileInfo is not null before calling.
        /// </summary>
        /// <param name="args"></param>
        /// <param name="extension"></param>
        /// <param name="info"></param>
        /// <param name="entry"></param>
        /// <param name="metadata"></param>
        private static void AddFileInfoToMap(Args args, FileExtension extension, ref AnalyzedInfo info, string entry, FileInfo metadata)
        {
            Trace.Assert(info.FileInfo != null);
            if (!info.FileInfo.TryGetValue(extension, out FileTypeInfo? value))
            {
                var file_type_info = new FileTypeInfo(0, new FileTypeInfoRecords(), new FileTypeInfoRecords("", ulong.MaxValue), 0, null, null);
                value = file_type_info;
                info.FileInfo.Add(extension, value);
            }

            value.NumFiles += 1;
            value.SizeInBytes += (ulong)metadata.Length;
            string? path = null;
            if (args.FullPath)
            {
                path = Path.GetFullPath(entry).TrimEnd(Path.AltDirectorySeparatorChar, Path.DirectorySeparatorChar);
            }
            else
            {
                path = entry;
            }
            if ((ulong)metadata.Length > value.LargestFile.Size)
            {
                value.LargestFile = new FileTypeInfoRecords(path, (ulong)metadata.Length);
            }
            if ((ulong)metadata.Length < value.SmallestFile.Size)
            {
                value.SmallestFile = new FileTypeInfoRecords(path, (ulong)metadata.Length);
            }

        }

        private static List<string>? SetUpIgnoreThese(Args args)
        {
            List<string>? ignore_these = null;
            if (args.IgnoreEntries != null)
            {
                ignore_these = new List<string>(args.IgnoreEntries.Split(',').Select(s => s.Trim()));
                foreach (var p in ignore_these)
                {
                    if (!Path.Exists(p))
                    {
                        Console.SetOut(Console.Error);
                        Console.WriteLine("WARNING: Can't ignore \"{0}\" because it doesn't exist", p);
                        Console.SetOut(Console.Out);
                    }
                }
            }
            return ignore_these;
        }

        private static Stopwatch? SetUpTimer(Args args)
        {
            if (args.Updates != null)
            {
                var timer = new Stopwatch();
                timer.Start();
                return timer;
            }

            return null;
        }

        /// <summary>
        /// Configures AnalyzedInfo based on the Args given
        /// </summary>
        /// <param name="args"></param>
        /// <returns></returns>
        private static AnalyzedInfo SetUpAnalyzedInfo(Args args)
        {
            AnalyzedInfo res = new() { FoundFiles = 0, FoundDirs = 0, TotalBytes = 0, FoundSymlinks = null, FileInfo = null };
            if (args.FileInfo)
            {
                res.FileInfo = [];
            }
            if (args.CountSymlinks)
            {
                res.FoundSymlinks = new SymlinkInfo();
            }
            return res;
        }
    }
}